File: docs/code-examples/arrow3d_simple_v2.rs
//! Log a batch of 3D arrows.

use std::f32::consts::TAU;

// TODO(cmc): use `rerun::components::Arrow3D` once migration is done.
use rerun::external::re_types::components::Arrow3D;
use rerun::{
    archetypes::Arrows3D, components::Color, datatypes::Vec3D, MsgSender, RecordingStreamBuilder,
};

fn main() -> Result<(), Box<dyn std::error::Error>> {
    let (rec_stream, storage) = RecordingStreamBuilder::new("arrow").memory()?;

    let (arrows, colors): (Vec<_>, Vec<_>) = (0..100)
        .map(|i| {
            let angle = TAU * i as f32 * 0.01;
            (
                {
                    let length = ((i + 1) as f32).log2();
                    Arrow3D::new(
                        Vec3D::ZERO,
                        [length * angle.sin(), 0.0, length * angle.cos()],
                    )
                },
                {
                    let c = (angle / TAU * 255.0) as u8;
                    Color::from_unmultiplied_rgba(255 - c, c, 128, 128)
                },
            )
        })
        .unzip();

    MsgSender::from_archetype("arrows", &Arrows3D::new(arrows).with_colors(colors))?
        .send(&rec_stream)?;

    rerun::native_viewer::show(storage.take())?;
    Ok(())
}

File: docs/code-examples/arrow3d_simple_v2.cpp
// Log a batch of 3D arrows.

#include <rerun.hpp>

#include <cmath>
#include <numeric>

namespace rr = rerun;

int main() {
    auto rr_stream = rr::RecordingStream("arrow3d");
    rr_stream.connect("127.0.0.1:9876");

    std::vector<rr::components::Vector3D> vectors;
    std::vector<rr::components::Color> colors;

    for (int i = 0; i < 100; ++i) {
        float angle = 2.0 * M_PI * i * 0.01f;
        float length = log2f(i + 1);
        vectors.push_back(rr::datatypes::Vec3D{length * sinf(angle), 0.0, length * cosf(angle)});

        // TODO(andreas): provide `unmultiplied_rgba`
        uint8_t c = static_cast<uint8_t>((angle / (2.0 * M_PI) * 255.0) + 0.5);
        uint32_t color = ((255 - c) << 24) + (c << 16) + (128 << 8) + (128 << 0);
        colors.push_back(color);
    }

    rr_stream.log("arrows", rr::archetypes::Arrows3D(vectors).with_colors(colors));
}
